File: api/app/src/models/PlaidItem.scala
package app.models

import java.time.Instant
import java.util.UUID

enum PlaidItemStatus:
  case Good, Bad, Unknown

  override def toString(): String =
    this match
      case Good    => "good"
      case Bad     => "bad"
      case Unknown => "unknown"

object PlaidItemStatus:
  def fromString(s: String): PlaidItemStatus =
    s match
      case "good"    => PlaidItemStatus.Good
      case "bad"     => PlaidItemStatus.Bad
      case "unknown" => PlaidItemStatus.Unknown

case class PlaidItem(
    id: UUID,
    userId: UUID,
    plaidAccessToken: String,
    plaidItemId: String,
    plaidInstitutionId: String,
    status: PlaidItemStatus,
    transactionsCursor: Option[String],
    createdAt: Instant,
    lastSyncedAt: Option[Instant],
    lastSyncError: Option[String],
    lastSyncErrorAt: Option[Instant],
    retryCount: Int
)

case class PlaidItemWithAccounts(
    plaidItem: PlaidItem,
    accounts: List[Account]
)

File: mobile/lib/src/connections/plaid_item.dart
import 'package:finny/src/accounts/account.dart';
import 'package:finny/src/institutions/institutions_model.dart';

class PlaidItem {
  final String id;
  final String institutionId;
  final String status;
  final DateTime createdAt;
  final DateTime? lastSyncedAt;
  final String? lastSyncError;
  final DateTime? lastSyncErrorAt;
  final int retryCount;
  List<Account> accounts;

  PlaidItem({
    required this.id,
    required this.institutionId,
    required this.status,
    required this.createdAt,
    this.lastSyncedAt,
    this.lastSyncError,
    this.lastSyncErrorAt,
    required this.retryCount,
    required this.accounts,
  });

  String get institutionName => institutions[institutionId]?.name ?? 'Unknown';

  @override
  String toString() {
    return 'PlaidItem{id: $id, institutionId: $institutionId, status: $status, createdAt: $createdAt, lastSyncedAt: $lastSyncedAt, lastSyncError: $lastSyncError, lastSyncErrorAt: $lastSyncErrorAt, retryCount: $retryCount, accounts: $accounts}';
  }
}
